mephisto/abstractions/database.py

```diff
@@ -92,6 +92,22 @@ def delete_qualification(self, qualification_name: str) -> None:
             provider = ProviderClass(self)
             provider.cleanup_qualification(qualification_name)
 
+    def optimized_load(
+        self,
+        target_cls,
+        db_id: str,
+        row: Optional[Mapping[str, Any]] = None,
+    ):
+        """
+        Load the given class in an optimized fashion, if this DB has a more
+        efficient way of storing and managing the data
+        """
+        return None
+
+    def cache_result(self, target_cls, value) -> None:
+        """Opportunity to store the result class from a load"""
+        return None
+
```

Comment:
I wonder if adding these optimization overrides to the abstract class may make it too specific/be a leaky abstraction. It seems like from the perspective of an abstract class, it doesn't matter if a "load" is optimized or not, a load is just a load.

I'm not sure of the full architecture here, but maybe another route could be the ability to create an OptimizedSQLLiteDatabase that inherits from a SQLLiteDatabase class which then overrides certain getters, setters, and loaders. And then you could use the optimized class instead. Just a thought.

Reply:
The difficulty here is that we want to be able to optimize the load with our most basic API - direct initialization (`Agent(db, db_id)`). I'm not sure how to expose such a possibility to the DB other than defining a method in the abstract base class. I suppose I could be more exact here in noting that this would be the approach to do a `possibly_load_from_cache`, as this is the only optimization form we have at the moment, but I'm not sure if another DB may do something differently in the future. In this case it's less the SQLite load that's a problem as much as the object initialization at this point.